File: engines/unity/flutter_plugin/android/src/main/kotlin/org/cytoid/gamecore/RuntimeState.kt
package org.cytoid.gamecore

/**
 * Native runtime lifecycle state per v2 host protocol § "Native Runtime Contract".
 *
 * Order matters: it is the order tests assert against and the order the iOS
 * Swift enum mirrors byte-for-byte. Do not re-order without updating
 * `RuntimeState.swift`, `RuntimeStateTest.kt`, and `RuntimeStateTests.swift`.
 *
 * The wire form is the lower-case enum name (e.g. `READY → "ready"`).
 */
enum class RuntimeState {
    UNAVAILABLE,
    STARTING,
    READY,
    BUSY,
    SUSPENDED,
    FAILED;

    /** v2 wire form, e.g. `READY → "ready"`. */
    val wireName: String get() = name.lowercase()

    companion object {
        /** Parse the wire form back to the enum, or null if unknown. */
        fun fromWireName(value: String?): RuntimeState? {
            if (value == null) return null
            return entries.firstOrNull { it.wireName == value }
        }
    }
}

File: engines/unity/flutter_plugin/ios/cytoid_game_core/Sources/cytoid_game_core/RuntimeState.swift
import Foundation

/// Native runtime lifecycle state per v2 host protocol § "Native Runtime Contract".
///
/// Order MUST match the Kotlin enum byte-for-byte — the cross-platform test
/// matrix (RuntimeStateTests.swift / RuntimeStateTest.kt) asserts the same
/// 6 states in the same order. Do not re-order without updating both files.
public enum RuntimeState: String, CaseIterable {
    case unavailable
    case starting
    case ready
    case busy
    case suspended
    case failed

    /// v2 wire form (already lowercased by the raw value).
    public var wireName: String { rawValue }

    /// Parse the wire form back to the enum, or nil if unknown.
    public static func from(wireName: String?) -> RuntimeState? {
        guard let wireName else { return nil }
        return RuntimeState(rawValue: wireName)
    }
}
